### PR TITLE
Refactor LLVM_OPT error check

### PR DESCRIPTION
**Pull request by saschanaz**
~~And remove some skips which are not really related with fastcomp version.~~ cc: @kripken

### CODE REVIEW

**Comment by saschanaz** [comment]
Those skipped tests run well on my Windows but not on CI and my Ubuntu environment...

**Comment by saschanaz** [comment]
Now I remember why I skipped it. `stddef.h` is not available for SDK-built-in clang by default, maybe it requires local clang installation? Canceled a commit that removed related skips.

**Comment by saschanaz** [comment]
The test pass with local clang installation, not sure what's wrong with the built-in one.

**Comment by saschanaz** [comment]
Installing clang from source works, so it seems precompiled SDK is a problem.

```
$ python2 /mnt/c/Users/sasch/Documents/GitHub/emsdk/emscripten/incoming/tests/runner.py other.tes
t_native_link_error_message
INFO:root:generating system asset: is_vanilla.txt... (this will be cached in "/home/saschanaz/.emscripten_cache/is_vanilla.txt" for subsequent builds)
INFO:root: - ok
WARNING:root:use EM_ALL_ENGINES=1 in the env to run against all JS engines, which is slower but provides more coverage
Test suites:
['test_other']
Running test_other: (1 tests)
test_native_link_error_message (test_other.other) ... clang version 4.0.0  (emscripten 1.37.25 : 1.37.25)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/saschanaz/emsdk-portable/clang/tag-e1.37.25/build_tag-e1.37.25_64/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5.4.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.0.0
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5.4.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
clang-4.0: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]
 "/home/saschanaz/emsdk-portable/clang/tag-e1.37.25/build_tag-e1.37.25_64/bin/clang-4.0" -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all -disable-free -disable-llvm-verifier -discard-value-names -main-file-name hello_world.c -mrelocation-model static -mthread-model posix -mdisable-fp-elim -fmath-errno -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu x86-64 -v -dwarf-column-info -debugger-tuning=gdb -coverage-notes-file /tmp/emscripten_test_other_ThL2C1/hello_world.gcno -resource-dir /home/saschanaz/emsdk-portable/clang/tag-e1.37.25/build_tag-e1.37.25_64/bin/../lib/clang/4.0.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/x86_64-linux-gnu/c++/5.4.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/x86_64-linux-gnu/c++/5.4.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/backward -internal-isystem /usr/local/include -internal-isystem /home/saschanaz/emsdk-portable/clang/tag-e1.37.25/build_tag-e1.37.25_64/bin/../lib/clang/4.0.0/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -fdeprecated-macro -fdebug-compilation-dir /tmp/emscripten_test_other_ThL2C1 -ferror-limit 19 -fmessage-length 120 -fobjc-runtime=gcc -fcxx-exceptions -fexceptions -fdiagnostics-show-option -fcolor-diagnostics -o hello_world.o -x c++ /mnt/c/Users/sasch/Documents/GitHub/emsdk/emscripten/incoming/tests/hello_world.c
clang -cc1 version 4.0.0 based upon LLVM 4.0.0 default target x86_64-unknown-linux-gnu
ignoring nonexistent directory "/include"
ignoring duplicate directory "/usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/x86_64-linux-gnu/c++/5.4.0"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0
 /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/x86_64-linux-gnu/c++/5.4.0
 /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/backward
 /usr/local/include
 /home/saschanaz/emsdk-portable/clang/tag-e1.37.25/build_tag-e1.37.25_64/bin/../lib/clang/4.0.0/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
/home/saschanaz/emsdk-portable/clang/tag-e1.37.25/build_tag-e1.37.25_64/bin/llvm-ar: creating hello_world.a
ok

----------------------------------------------------------------------
Ran 1 test in 3.142s

OK
```

**Comment by kripken** [comment]
Hmm, on the other hand it's better to check that command. Perhaps we should modify the test and the behavior it checks? @juj, you wrote those.

Maybe we could catch a thrown exception, add the extra logging for that test, then rethrow, something like that.

**Comment by saschanaz** [comment]
This native header failure also affects every SSE test, I think the precompiled binaries probably should bundle their own headers (as they are not that big).

**Comment by saschanaz** [comment]
Currently this PR doesn't fix anything, so closing. Related: #5618

**Comment by saschanaz** [comment]
Oops, it does fix `other.test_native_link_error_message` but just not on CI.

**Comment by kripken** [comment]
I'm still concerned about this - we shouldn't weaken general checking, we should try to find something to do with that test (fix it, or add a check to skip it on systems where it fails).

**Comment by saschanaz** [comment]
I found that @juj already fixed this on #6044 but I tried a slight refactoring by removing `check=False` and catching exceptions. Looks better to me, what do you think? @kripken

File: tools/shared.py [hunk 1]
```diff
@@ -1873,10 +1873,11 @@ def llvm_opt(filename, opts, out=None):
...
     logging.debug('emcc: LLVM opts: ' + ' '.join(opts) + '  [num inputs: ' + str(len(inputs)) + ']')
     target = out or (filename + '.opt.bc')
-    proc = run_process([LLVM_OPT] + inputs + opts + ['-o', target], stdout=PIPE, check=False)
-    output = proc.stdout
-    if proc.returncode != 0 or not os.path.exists(target):
-      logging.error('Failed to run llvm optimizations: ' + output)
+    try:
+      run_process([LLVM_OPT] + inputs + opts + ['-o', target], stdout=PIPE)
+      assert os.path.exists(target), 'llvm optimizer emitted no output.'
+    except CalledProcessError as e:
```
Comment (kripken) [inline tools/shared.py]: doesn't this need `subprocess.CalledProcessError`?

**Comment by kripken** [comment]
Aside from that question, I don't have strong feelings either way, both approaches seem reasonable.

File: tools/shared.py (hunk 1 above)
Comment (saschanaz) [reply tools/shared.py]: Oops, you're right.

**Comment by juj** [comment]
lgtm, mostly due to dropping the redundant looking `output = proc.stdout` line, and nice to also have fewer boolean extra checks around.

**Pull request merged**